compiler/rustc_expand/src/expand.rs

```diff
@@ -1299,36 +1307,37 @@ impl<'a, 'b> MutVisitor for InvocationCollector<'a, 'b> {
 
                         *mod_kind = ModKind::Loaded(krate.items, Inline::No, inner_span);
                         item.attrs = krate.attrs;
-                        // File can have inline attributes, e.g., `#![cfg(...)]` & co. => Reconfigure.
-                        item = match self.configure(item) {
-                            Some(node) => node,
-                            None => {
-                                if *pushed {
-                                    sess.included_mod_stack.borrow_mut().pop();
-                                }
-                                return Default::default();
-                            }
-                        };
                         dir
                     }
                 };
 
-                // Set the module info before we flat map.
-                self.cx.current_expansion.directory_ownership = ownership;
-                module.directory = path;
-                module.mod_path.push(ident);
-                let orig_module =
-                    mem::replace(&mut self.cx.current_expansion.module, Rc::new(module));
-
-                let result = noop_flat_map_item(item, self);
+                if item.attrs.len() > old_attrs_len {
+                    // If we loaded an out-of-line module and added some inner attributes,
+                    // then we need to re-configure it and try to re-collect attributes
+                    // for resolution and expansion, so we are calling `flat_map_item` again.
+                    if *pushed {
+                        sess.included_mod_stack.borrow_mut().pop();
+                    }
 
-                // Restore the module info.
-                self.cx.current_expansion.module = orig_module;
-                self.cx.current_expansion.directory_ownership = orig_ownership;
-                if *pushed {
-                    sess.included_mod_stack.borrow_mut().pop();
+                    self.flat_map_item(item)
```

Comment:
This line was a little confusing to me. If I understand correctly, we should recurse at most once - the second time around, we will have `ModKind::Loaded` and therefore never add any new attributes. Could you add a comment explaining why this recursion is ok?

Reply:
I've removed the recursion and now call configuration and attribute collection directly.
This also allows to avoid re-calculating module paths and ownership in the common case of no collected custom inner attributes.